iwyu.cc

```diff
@@ -1447,21 +1447,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
-  set<const Type*> GetProvidedTypesForAlias(const TypedefNameDecl* decl) const {
-    const Type* underlying_type = decl->getUnderlyingType().getTypePtr();
+  set<const Type*> GetProvidedTypesForAlias(const Type* underlying_type,
+                                            SourceLocation decl_loc) const {
     // If the underlying type is itself a typedef, we recurse.
     if (const auto* underlying_typedef =
             underlying_type->getAs<TypedefType>()) {
       if (const auto* underlying_typedef_decl = dyn_cast<TypedefNameDecl>(TypeToDeclAsWritten(underlying_typedef))) {
         // TODO(csilvers): if one of the intermediate typedefs
         // #includes the necessary definition of the 'final'
         // underlying type, do we want to return it here?
-        return GetProvidedTypesForAlias(underlying_typedef_decl);
+        return GetProvidedTypesForAlias(
+            underlying_typedef_decl->getUnderlyingType().getTypePtr(),
```

Comment:
Wait, are we now going from type to decl and then back to type? Is there no way to get from `underlying_typedef` to the underlying type directly? OTOH, it seems like we need the decl location.

Reply:
`underlying_typedef->desugar().getTypePtr()` would probably work. However, the approach under discussion was already present on the master branch [in the calling code](https://github.com/include-what-you-use/include-what-you-use/blob/0.21/iwyu.cc#L2607) (inside `ReportTypeUseInternal`). I think, if it should be changed here, it should also be changed there. But it is not obvious to me why one approach is better than the other, if the `TypedefNameDecl` is still needed for other purposes.

Reply:
It looked very roundabout, but as you say, it makes sense if the decl is needed anyway (and I suppose it is, both to determine if we should report, and for the location).

Reply:
To be more correct, its location is taken to determine if a particular type component should be reported.